src/server/editorServices.ts

```diff
@@ -760,6 +761,7 @@ namespace ts.server {
         readonly toCanonicalFileName: (f: string) => string;
 
         public readonly host: ServerHost;
+        public readonly fshost: ServerHost | undefined; // TODOO: Should be TestFSWithWatch.VirtualServerHost | undefined

```

Comment:
That works, except for two things:

1. modifyFile and ensureFileOrFolder are vfs-specific. I'll intersect those from VirtualServerHost.
2. getOrCreateScriptInfoWorker calls `new ScriptInfo` with `this.fshost || this.host`, which requires 14 methods not otherwise required. I think I'll try pass both to ScriptInfo in this case.

Reply:
1. It turns out be more efficient and safer to *not* intersect anything from VirtualServerHost and just case in the new updateFileSystem code.
2. I modified ScriptInfo and its dependencies to use FSHost, which only required a couple more methods.